sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+			AmountPaidMsat: int64(amtPaid),
+		},
+	)
+
+	return err
+}
+
+// UpdateAmpState updates the state of the AMP sub invoice identified by the
+// setID.
+func (s *sqlInvoiceUpdater) UpdateAmpState(setID [32]byte,
+	newState invpkg.InvoiceStateAMP, _ models.CircuitKey) error {
+
+	var (
+		settledIndex sql.NullInt64
+		settledAt    sql.NullTime
+	)
+
+	switch newState.State {
+	case invpkg.HtlcStateSettled:
+		nextSettledIndex, err := s.db.NextInvoiceSettledIndex(s.ctx)
+		if err != nil {
+			return err
+		}
+
+		settledIndex = sqlInt64(nextSettledIndex)
+
+		// If the invoice is settled, we'll also update the settle time.
+		settledAt = sql.NullTime{
+			Time:  s.updateTime.UTC(),
+			Valid: true,
+		}
+
+		err = s.db.OnAMPSubInvoiceSettled(
+			s.ctx, sqlc.OnAMPSubInvoiceSettledParams{
+				AddedAt:   s.updateTime.UTC(),
+				InvoiceID: int64(s.invoice.AddIndex),
+				SetID:     setID[:],
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+	case invpkg.HtlcStateCanceled:
+		err := s.db.OnAMPSubInvoiceCanceled(
+			s.ctx, sqlc.OnAMPSubInvoiceCanceledParams{
+				AddedAt:   s.updateTime.UTC(),
+				InvoiceID: int64(s.invoice.AddIndex),
+				SetID:     setID[:],
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := s.db.UpdateAMPSubInvoiceState(
+		s.ctx, sqlc.UpdateAMPSubInvoiceStateParams{
+			SetID:        setID[:],
+			State:        int16(newState.State),
+			SettledIndex: settledIndex,
+			SettledAt:    settledAt,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Commit finalizes the update before it is written to the database. Note that
+// we don't use this directly in the SQL implementation, so the function is just
+// a stub.
+func (s *sqlInvoiceUpdater) Commit(_ invpkg.UpdateType) error {
+	return nil
+}
+
+// UpdateInvoice attempts to update an invoice corresponding to the passed
+// payment hash. If an invoice matching the passed payment hash doesn't exist
+// within the database, then the action will fail with a "not found" error.
+//
+// The update is performed inside the same database transaction that fetches the
+// invoice and is therefore atomic. The fields to update are controlled by the
+// supplied callback.
+func (i *InvoiceStore) UpdateInvoice(ctx context.Context, ref invpkg.InvoiceRef,
+	_ *invpkg.SetID, callback invpkg.InvoiceUpdateCallback) (
+	*invpkg.Invoice, error) {
+
+	var updatedInvoice *invpkg.Invoice
+
+	txOpt := InvoiceQueriesTxOptions{readOnly: false}
+	txErr := i.db.ExecTx(ctx, &txOpt, func(db InvoiceQueries) error {
+		invoice, err := i.fetchInvoice(ctx, db, ref)
+		if err != nil {
+			return err
+		}
+
+		updateTime := i.clock.Now()
+		updater := &sqlInvoiceUpdater{
+			db:         db,
+			ctx:        ctx,
+			invoice:    invoice,
+			updateTime: updateTime,
+		}
+
+		payHash := ref.PayHash()
+		updatedInvoice, err = invpkg.UpdateInvoice(
+			payHash, invoice, updateTime, callback, updater,
+		)
+
+		return err
+	})
+	if txErr != nil {
+		return updatedInvoice, txErr
```

Comment:
I wonder why we are not returning `nil` here when we have a `txErr`

Reply:
Good catch, added clarification here.